### PR TITLE
[HW] Add support for vfrsqrt7 instruction

### PR DESCRIPTION
**Pull request by Shafi10x**
vfrsqrt7 is a unary vector-vector instruction that returns an estimate of 1/sqrt(x) accurate to 7 bit.
### Input
For the non-exceptional cases:
- If input is normal, normalized input exponent is equal to the input exponent and normalized input significand(mantissa) is equal to the input significand.
- If input is subnormal, normalized input exponent is equal to 0 minus the number of leading zeros in the signifcand and the normalized input significand is given by shifting the input significand left by 1 minus the normalized input exponent.
### Output
- Output exponent can be found by:
                exp_o = (3*B-1-exp_i )/2
                         where B is exponent bias.
- Output 7-bit significand can be found by using lookup table.
- The address of lookup table is found by concatenating last LSB of the normalized input exponent and the six MSBs of the normalized input significand.
- The remainder of the result significand is zero
- The output sign equals the input sign.
### Fixed

- N/A

### Added

- Added support for vfrsqrt7 instruction
- Added tests for vfrsqrt7 instruction

### Changed

- N/A

## Checklist

- [x] Automated tests pass
- [x] Changelog updated
- [x] Code style guideline is observed

Please check our [contributing guidelines](CONTRIBUTING.md) before opening a Pull Request.

### CODE REVIEW

File: hardware/src/lane/vmfpu.sv
Comment (mp-17) [inline hardware/src/lane/vmfpu.sv:882]: Why are you delaying `operand_a` by one cycle? Is it because the `FPU` latency is 1 cycle for the `CLASSIFY` operation, and you need the source operand as well? 
If so, it would be better to parametrize this operation on the `LatFNonComp` parameter so that this instruction does not break when the parameter is tuned. You can have a look at https://github.com/pulp-platform/ara/blob/main/hardware/src/lane/simd_mul.sv#L71 for the "pipeline" creation.
This can keep full throughput for the operation, otherwise, you can also keep only one register to delay `operand_a` by 1 cycle regardless of the latency of the `CLASSIFY` op and simply mask the FPU input ready until you are not ready to buffer another `operand_a`, i.e. when you are over with the current result.

File: hardware/src/lane/vmfpu.sv
Comment (mp-17) [inline hardware/src/lane/vmfpu.sv:938]:
```verilog
for (int h = 0; h < 4; h++) vfrsqrt7_out_e16[h] = 
  vfrsqrt7_fp16(vfpu_result[h*16 +: 10], operand_a_delay[h*16 +: 16], lzc_e16[h*4 +: 4]);
```

File: hardware/src/lane/vmfpu.sv
Comment (mp-17) [inline hardware/src/lane/vmfpu.sv:949]:
```verilog
for (int w = 0; w < 2; w++) vfrsqrt7_out_e32[w] = 
  vfrsqrt7_fp32(vfpu_result[w*32 +: 10], operand_a_delay[w*32 +: 32], lzc_e32[w*5 +: 5]);
```

File: hardware/src/lane/vmfpu.sv
Comment (mp-17) [inline hardware/src/lane/vmfpu.sv:957]:
```verilog
for (int d = 0; d < 1; d++) vfrsqrt7_out_e64[d] = 
  vfrsqrt7_fp64(vfpu_result[d*64 +: 10], operand_a_delay[d*64 +: 64], lzc_e64[d*6 +: 6]);
```

**Review by mp-17 — CHANGES_REQUESTED** [review]
Thanks a lot for the contribution! I started looking at it and writing some comments while the tool is synthesizing and implementing.
Then, I will check the mathematical part with more attention.

File: hardware/src/lane/vmfpu.sv
Comment (Shafi10x) [reply hardware/src/lane/vmfpu.sv:882]: Should I need to change `vfpu_flag_mask` also?

File: hardware/include/ara_pkg.sv
Comment (Shafi10x) [reply hardware/include/ara_pkg.sv:1005]: As it was used previously in the code that's why I used the same thing. I can change it `logic [31:0]` if you want.

File: apps/riscv-tests/isa/rv64uv/vfrsqrt7.c
Comment (Shafi10x) [reply apps/riscv-tests/isa/rv64uv/vfrsqrt7.c:5]: Sure. :-)

File: hardware/include/ara_pkg.sv
Comment (Shafi10x) [reply hardware/include/ara_pkg.sv:1052]: Yes.

**Comment by Shafi10x** [comment]
@mp-17 I have updated all the changes according to your suggestions.

**Comment by mp-17** [comment]
Thanks a lot and sorry for the delay! I am checking rn!

**Review by mp-17 — CHANGES_REQUESTED** [review]
Okay, thanks a lot! It looks good; let's see the CI. 
Now, I will just 1) make some style changes and 2) parameterize these more "exotic" floating-point instructions since they are not optimized inside the FPU. I will then issue a PR on your `10x` fork with my changes. After the iteration, we can merge.

File: hardware/include/ara_pkg.sv
Comment (Shafi10x) [reply hardware/include/ara_pkg.sv]: `vfrsqrt7_exp_i` is mux output, I mean it is not awlays `vfrsqrt7_exp_i = 12'd0 - {6'd0,leading_zeros_count};`. So, `vfrsqrt7_exp_o = E64_3xB + leading_zeros_count - 1;` is not possible.

**Comment by Shafi10x** [comment]
>

File: hardware/include/ara_pkg.sv
Comment (Shafi10x) [reply hardware/include/ara_pkg.sv]: I have also parametrized `vfpu_flag_mask` . I have tested `vfrsqrt7` and ` vfclass` both by changing  the parameter `LatFNonComp` value and it works fine.

Comment (mp-17) [reply hardware/include/ara_pkg.sv]: Thanks a lot! 

> `vfrsqrt7_exp_i` is mux output, I mean it is not awlays `vfrsqrt7_exp_i = 12'd0 - {6'd0,leading_zeros_count};`. So, `vfrsqrt7_exp_o = E64_3xB + leading_zeros_count - 1;` is not possible.

It should be possible by controlling the output mux with both `vfpu_result[6:5]` and `vfpu_result`, like

```verilog
case ({vfpu_result[6:5], vfpu_result}) 
  {2'b01, fpnew_pkg::POSNORM}: begin
    ...
  end
endcase
```

But I don't know if this is needed. And if we parametrize the support for this instruction, we can optimize it afterward with no problems.

**Comment by mp-17** [comment]
Let's wait for the CI, then! :-)

File: hardware/src/ara_dispatcher.sv
Comment (Shafi10x) [inline hardware/src/ara_dispatcher.sv:3025]: Should this comment be like  `Check that we have vfrsqrt7 support..`?

Comment (mp-17) [reply hardware/src/ara_dispatcher.sv:3025]: Definitely, thanks! :-)

**Comment by mp-17** [comment]
I see there is a problem with the `riscv-tests-simv` job. Once it is solved and the branch is rebased, we can merge!

**Comment by Shafi10x** [comment]
Can I get permission to run CI checks on this PR?

**Comment by Shafi10x** [comment]
@mp-17  The CI checks have passed and all conflicts have resolved. Can we merge this?

**Comment by mp-17** [comment]
Yes, definitely! Can you please rebase? I will merge afterwards :-)

**Comment by mp-17** [comment]
I still see there are conflicts. Have you rebased on the current `main`? 
(sorry for this rebase ping-pong!)

**Comment by Shafi10x** [comment]
@mp-17 Yesterday, I have resolved all the conflicts. I am not seeing any conflicts right now.

**Comment by mp-17** [comment]
I think you merged our `main` into the branch instead of rebasing it on top of `main`.

**Comment by Shafi10x** [comment]
@mp-17 now this PR is ready for merging.

**Comment by Shafi10x** [comment]
@mp-17 Can we merge this PR?

**Comment by mp-17** [comment]
Done, thank you again.

**Pull request merged**
